File: src/main/kotlin/com/mdreamfever/netty/example/echo/EchoClient.kt
package com.mdreamfever.netty.example.echo

import io.netty.bootstrap.Bootstrap
import io.netty.channel.ChannelInitializer
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.SocketChannel
import io.netty.channel.socket.nio.NioSocketChannel
import io.netty.handler.codec.string.StringDecoder
import io.netty.handler.codec.string.StringEncoder
import java.net.InetSocketAddress

fun main() {
    val client = EchoClient()
    client.connect()
}

class EchoClient(private val inetAddress: InetSocketAddress = InetSocketAddress("127.0.0.1", 8089)) {
    private val bootstrap = Bootstrap()
    private val group = NioEventLoopGroup()

    init {
        bootstrap
            .group(group)
            .remoteAddress(inetAddress)
            .channel(NioSocketChannel::class.java)
            .handler(object : ChannelInitializer<SocketChannel>() {
                override fun initChannel(ch: SocketChannel) {
                    ch.pipeline().run {
                        addLast(StringDecoder())
                        addLast(StringEncoder())
                        addLast(EchoClientHandler())
                    }
                }
            })
    }

    fun connect() {
        try {
            val channel = bootstrap.connect().sync().channel()
            channel.writeAndFlush("hello world")
            channel.closeFuture().sync()
        } finally {
            group.shutdownGracefully()
        }
    }
}

File: src/main/scala/echo/EchoClient.scala
package echo

import io.netty.bootstrap.Bootstrap
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.nio.NioSocketChannel
import io.netty.handler.ssl.SslContextBuilder
import io.netty.handler.ssl.util.InsecureTrustManagerFactory

import java.net.InetAddress

object EchoClient {
  private val ssl = true
  val port: Int = if (ssl) 8443 else 8989
  private val bootstrap = new Bootstrap()
  private val workGroup = new NioEventLoopGroup()
  private val sslContext = if (ssl) SslContextBuilder.forClient().trustManager(InsecureTrustManagerFactory.INSTANCE).build() else null
  bootstrap.group(workGroup)
    .remoteAddress(InetAddress.getLocalHost, port)
    .channel(classOf[NioSocketChannel])
    .handler(new EchoClientChannelInitializer(sslContext))

  def main(args: Array[String]): Unit = {
    try {
      val channel = bootstrap.connect().sync().channel()
      channel.writeAndFlush("hello world")
      channel.closeFuture().sync()
    } finally {
      workGroup.shutdownGracefully()
    }
  }
}
